common/cache/domainCache.go

```diff
@@ -185,13 +188,13 @@ func (c *domainCache) GetAllDomain() map[string]*DomainCacheEntry {
 // afterCallback will be invoked when NOT holding the domain cache lock.
 func (c *domainCache) RegisterDomainChangeCallback(shard int, initialNotificationVersion int64, beforeCallback CallbackFn, afterCallback CallbackFn) {
 	c.Lock()
-	defer c.Unlock()
-
 	c.beforeCallbacks[shard] = beforeCallback
 	c.afterCallbacks[shard] = afterCallback
+	domainNotificationVersion := c.domainNotificationVersion
+	c.Unlock()
```

Comment:
Looks like both before and after callbacks were called within the lock before this change and now they are called outside of the lock.  Do we even need 2 separate callbacks?

Reply:
we do not really need a 2 type of callbacks, one before one after, i thought providing 2 type of callbacks can be useful